lib/matplotlib/backends/backend_qtagg.py

```diff
@@ -71,6 +71,15 @@ def paintEvent(self, event):
         finally:
             painter.end()
 
+    def print_figure(self, *args, **kwargs):
+        super().print_figure(*args, **kwargs)
+        # In some cases, Qt will itself trigger a paint event after closing the file
+        # save dialog. When that happens, we need to be sure that the internal canvas is
+        # re-drawn. However, if the user is using an automatically-chosen Qt backend but
+        # saving with a different backend (such as pgf), we do not want to trigger a
+        # full draw in Qt, so just set the flag for next time.
```

Comment:
This is going to force a full re-render the next time the window paints itself no matter which canvas class was actually selected.

The advantage of doing this rather than `self.draw_idle8 that this won't _trigger_ the repaint and redraw on its own?

Reply:
> This is going to force a full re-render the next time the window paints itself no matter which canvas class was actually selected.

Yes, this would be the same as before #25713.

> The advantage of doing this rather than `self.draw_idle8 that this won't _trigger_ the repaint and redraw on its own?

Yes, `draw_idle` will add a timer and actually draw later; we don't want that in order to keep the fix for #5234.

Reply:
@QuLogic Thanks for investigating this.  On the side of mplcairo.qt I had the same issue, but took a slightly different approach to fix it (https://github.com/matplotlib/mplcairo/commit/cd6d2ca3c92d29196441bad37ad2fa68582af079): I stash the old renderer before calling print_figure, and then restore it at the end, which always saves the need for a redraw.  Perhaps the same can be done here?  (Not that I actually checked.)

Reply:
This suggests a more general change in that `print_figure` should _always_ make a new canvas object even if the currently in-use canvas can technically save the format requested.  This would simplify some of the logic a bit (I think) and would probably save a bunch of dpi changing work...